### PR TITLE
Fix: Immutable DirectRelationReference

### PR DESCRIPTION
**Pull request by doctrino**
## Description
Please describe the change you have made.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/client/data_classes/data_modeling/data_types.py [hunk 1]
```diff
@@ -18,20 +18,24 @@
...
     external_id: str
 
     def dump(self, camel_case: bool = True) -> dict[str, str | dict]:
-        output = asdict(self)
-
-        return convert_all_keys_recursive(output, camel_case)
+        return {
+            "space": self.space,
+            "externalId" if camel_case else "external_id": self.external_id,
+        }
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/data_types.py:30]: Increased efficiency.

File: cognite/client/data_classes/data_modeling/data_types.py [hunk 2]
```diff
@@ -18,20 +18,24 @@
...
 
     @classmethod
     def load(cls, data: dict | tuple[str, str]) -> DirectRelationReference:
         if isinstance(data, dict):
-            return cls(**convert_all_keys_to_snake_case(rename_and_exclude_keys(data, exclude={"type"})))
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/data_types.py:34]: This is a bug as it does not handle adding optional parameters in the API

File: cognite/client/data_classes/data_modeling/data_types.py [hunk 3]
```diff
@@ -18,20 +18,24 @@
...
+            return cls(
+                space=data["space"],
+                external_id=data["externalId"] if "externalId" in data else data["external_id"],
```
Comment (erlendvollset) [inline cognite/client/data_classes/data_modeling/data_types.py:37]: load shouldn't need to handle snake cade

File: cognite/client/data_classes/data_modeling/data_types.py (hunk 1 above)
Comment (haakonvt) [reply cognite/client/data_classes/data_modeling/data_types.py:30]: I like it, `asdict` is stupidly slow 😜

**Review by erlendvollset — APPROVED** [review]
LGTM!

**Pull request merged**